From 165bdf811042fb08f0b644f620fb6889733ddfa6 Mon Sep 17 00:00:00 2001
From: Robert Iannucci <iannucci@chromium.org>
Date: Wed, 6 Mar 2024 16:09:44 -0800
Subject: [PATCH] Make automake relocatable

This is the package definition for the automake tools and includes
a patch to make them relocatable (written against automake 1.15).

By default automake hard-codes the --prefix value into the binaries it
deploys, making them unsuitable for relocatable deployment (e.g. with
CIPD). The patch here replaces all the hard-coded paths with either:
  * The assumption that the tool is in $PATH, and so "/path/to/tool" is
    replaced by "tool"
  * The assumption that the data files are relative to the binary being
    run, e.g. if we're running ".../bin/tool" that we can find the data
    files at ".../share/extra_files".

The patch was made by doing `make install` on the base package, looking
for absolute paths and then changing the sources so the absolute paths
no longer showed up in the output of `make install`.
---
 bin/aclocal.in         | 14 ++++++++++----
 bin/automake.in        |  7 +++++--
 lib/Automake/Config.in |  3 ++-
 3 files changed, 17 insertions(+), 7 deletions(-)

diff --git a/bin/aclocal.in b/bin/aclocal.in
index f112447..e03b1d6 100644
--- a/bin/aclocal.in
+++ b/bin/aclocal.in
@@ -25,8 +25,11 @@ use warnings FATAL => 'all';
 
 BEGIN
 {
-  unshift (@INC, '@datadir@/@PACKAGE@-@APIVERSION@')
-    unless $ENV{AUTOMAKE_UNINSTALLED};
+  use File::Basename;
+  my $PREFIX = dirname(dirname(__FILE__));
+  @Aclocal::perl_libdirs = ("$PREFIX/share/@PACKAGE@-@APIVERSION@")
+    unless @Aclocal::perl_libdirs;
+  unshift @INC, @Aclocal::perl_libdirs;
 }
 
 use File::Basename;
@@ -40,6 +43,9 @@ use Automake::ChannelDefs;
 use Automake::XFile;
 use Automake::FileUtils;
 
+my $PREFIX = dirname(dirname(__FILE__));
+my $DATADIR = "$PREFIX/share";
+
 # Some globals.
 
 # Support AC_CONFIG_MACRO_DIRS also with older autoconf.
@@ -65,8 +71,8 @@ $perl_threads = 0;
 # ACLOCAL_PATH environment variable, and reset with the '--system-acdir'
 # option.
 my @user_includes = ();
-my @automake_includes = ('@datadir@/aclocal-' . $APIVERSION);
-my @system_includes = ('@datadir@/aclocal');
+my @automake_includes = ("$DATADIR/aclocal-$APIVERSION");
+my @system_includes = ("$DATADIR/aclocal");
 
 # Whether we should copy M4 file in $user_includes[0].
 my $install = 0;
diff --git a/bin/automake.in b/bin/automake.in
index 30babd6..70528be 100644
--- a/bin/automake.in
+++ b/bin/automake.in
@@ -28,8 +28,11 @@ use warnings FATAL => 'all';
 
 BEGIN
 {
-  unshift (@INC, '@datadir@/@PACKAGE@-@APIVERSION@')
-    unless $ENV{AUTOMAKE_UNINSTALLED};
+  use File::Basename;
+  my $PREFIX = dirname(dirname(__FILE__));
+  @Automake::perl_libdirs = ("$PREFIX/share/@PACKAGE@-@APIVERSION@")
+    unless @Automake::perl_libdirs;
+  unshift @INC, @Automake::perl_libdirs;
 
   # Override SHELL.  This is required on DJGPP so that system() uses
   # bash, not COMMAND.COM which doesn't quote arguments properly.
diff --git a/lib/Automake/Config.in b/lib/Automake/Config.in
index d529f1b..c4ecab9 100644
--- a/lib/Automake/Config.in
+++ b/lib/Automake/Config.in
@@ -22,6 +22,7 @@ use strict;
 use warnings FATAL => 'all';
 
 use Exporter;
+use File::Basename;
 
 our @ISA = qw (Exporter);
 our @EXPORT = qw ($APIVERSION $PACKAGE $PACKAGE_BUGREPORT $VERSION
@@ -34,7 +35,7 @@ our $PACKAGE = '@PACKAGE@';
 our $PACKAGE_BUGREPORT = '@PACKAGE_BUGREPORT@';
 our $VERSION = '@VERSION@';
 our $RELEASE_YEAR = '@RELEASE_YEAR@';
-our $libdir = $ENV{"AUTOMAKE_LIBDIR"} || '@datadir@/@PACKAGE@-@APIVERSION@';
+our $libdir = dirname(dirname(__FILE__));
 
 our $perl_threads = 0;
 # We need at least this version for CLONE support.
-- 
2.44.0.278.ge034bb2e1d-goog

